src/ckb/channel.rs

```diff
@@ -1751,25 +1925,163 @@ impl ChannelActorState {
+        } else {
+            (tlc.committed_at.get_remote(), tlc.committed_at.get_local())
+        };
+        let local_pubkey = derive_tlc_pubkey(
+            &self.get_local_channel_parameters().pubkeys.tlc_base_key,
+            &self.get_local_commitment_point(remote_commitment_number),
         );
-        debug!(
-            "Current pending received tlcs: {:?}",
-            self.pending_received_tlcs
+        let remote_pubkey = derive_tlc_pubkey(
+            &self.get_remote_channel_parameters().pubkeys.tlc_base_key,
+            &self.get_remote_commitment_point(local_commitment_number),
         );
-        self.pending_offered_tlcs
-            .values()
-            .chain(self.pending_received_tlcs.values())
-            .cloned()
+
+        if is_offered {
+            (local_pubkey, remote_pubkey)
+        } else {
+            (remote_pubkey, local_pubkey)
+        }
+    }
+
+    pub fn get_active_received_tlc_with_pubkeys(
+        &self,
+    ) -> impl Iterator<Item = (&DetailedTLCInfo, Pubkey, Pubkey)> {
+        self.get_active_received_tlcs().into_iter().map(move |tlc| {
+            let (k1, k2) = self.get_tlc_pubkeys(tlc);
+            (tlc, k1, k2)
+        })
+    }
+
+    pub fn get_active_offered_tlc_with_pubkeys(
+        &self,
+    ) -> impl Iterator<Item = (&DetailedTLCInfo, Pubkey, Pubkey)> {
+        self.get_active_offered_tlcs().into_iter().map(move |tlc| {
+            let (k1, k2) = self.get_tlc_pubkeys(tlc);
+            (tlc, k1, k2)
+        })
+    }
+
+    pub fn get_witness_args_for_active_tlcs(&self, local: bool) -> Vec<u8> {
+        // Build a sorted array of TLC so that both party can generate the same commitment transaction.
+        debug!("All tlcs: {:?}", self.tlcs);
+        let tlcs = {
+            let (mut received_tlcs, mut offered_tlcs) = (
+                self.get_active_received_tlc_with_pubkeys()
+                    .into_iter()
+                    .map(|(tlc, local, remote)| (tlc.clone(), local, remote))
+                    .collect::<Vec<_>>(),
+                self.get_active_offered_tlc_with_pubkeys()
+                    .into_iter()
+                    .map(|(tlc, local, remote)| (tlc.clone(), local, remote))
+                    .collect::<Vec<_>>(),
+            );
+            debug!("Received tlcs: {:?}", &received_tlcs);
+            debug!("Offered tlcs: {:?}", &offered_tlcs);
+            let (mut a, mut b) = if local {
+                (received_tlcs, offered_tlcs)
+            } else {
+                for tlc in received_tlcs.iter_mut().chain(offered_tlcs.iter_mut()) {
+                    // Need to flip these fields for the counterparty.
+                    tlc.0.tlc.flip_mut();
+                }
+                (offered_tlcs, received_tlcs)
+            };
+            a.sort_by(|x, y| u64::from(x.0.tlc.id).cmp(&u64::from(y.0.tlc.id)));
+            b.sort_by(|x, y| u64::from(x.0.tlc.id).cmp(&u64::from(y.0.tlc.id)));
```

Comment:
BTreeMap is sorted, I don't think we need sort again here. btw, we need unit test for this fn.